daml-lf/tests/MultiKeys.daml

```diff
@@ -0,0 +1,232 @@
+         archive (fromSome (find (/= keyCid) [cid1, cid2]))
+         None <- lookupByKey @Keyed p
+         pure ()
+
+    -- should be accepted in uck
+    nonconsuming choice GlobalArchiveOverwritesKnownGlobal2 : ()
+      with
+        cid1 : ContractId Keyed
+        cid2 : ContractId Keyed
+      controller p
+      do cid1 =/= cid2
+         c1 <- fetch cid1
+         key c1 === p
+         c2 <- fetch cid2
+         key c2 === p
+         (keyCid, _) <- fetchByKey @Keyed p
+         assert (keyCid `elem` [cid1, cid2])
+         archive keyCid
+         None <- lookupByKey @Keyed p
+         pure ()
+
+    -- should be rejected in uck
+    nonconsuming choice RollbackCreateNonRollbackFetchByKey : ()
+      controller p
+      do try do
+           create (Keyed p)
+           throw E
+         catch E -> pure ()
+         _ <- fetchByKey @Keyed p
+         pure ()
+
+    -- should be rejected in uck
+    nonconsuming choice RollbackFetchByKeyNonRollbackCreate : ()
+      controller p
+      do try do
+           _ <- fetchByKey @Keyed p
+           throw E
+         catch E -> pure ()
+         cid <- create (Keyed p)
+         (cid', _) <- fetchByKey @Keyed p
+         cid === cid'
+
+    -- should be accepted in uck
+    nonconsuming choice RollbackFetchNonRollbackCreate : ()
+      with
+        cid : ContractId Keyed
+      controller p
+      do try do
+           c <- fetch cid
+           key c === p
+           throw E
+         catch E -> pure ()
+         cid <- create (Keyed p)
+         (cid', _) <- fetchByKey @Keyed p
+         cid === cid'
+
+    -- should be rejected in uck
+    nonconsuming choice RollbackGlobalArchiveNonRollbackCreate : ()
+      with
+        cid : ContractId Keyed
+      controller p
+      do try do
+           c <- fetch cid
+           key c === p
+           -- Again, consuming choices are implicitly by-key.
+           archive cid
+           throw E
+         catch E -> pure ()
+         cid <- create (Keyed p)
+         (cid', _) <- fetchByKey @Keyed p
+         cid === cid'
+
+    -- should be rejected in uck
+    nonconsuming choice RollbackCreateNonRollbackGlobalArchive : ()
+      with
+        cid : ContractId Keyed
+      controller p
+      do try do
+           create (Keyed p)
+           throw E
+         catch E -> pure ()
+         c <- fetch cid
+         key c === p
+         archive cid
+         None <- lookupByKey @Keyed p
+         pure ()
+
+    -- should be rejected in uck
+    nonconsuming choice RollbackGlobalArchiveUpdates : ()
+      with
+        cid1 : ContractId Keyed
+        cid2 : ContractId Keyed
+      controller p
+      do (cid', _) <- fetchByKey @Keyed p
+         cid1 === cid'
+         try do
+           -- archive has an implicit read which is not rolled back
+           -- and should therefore update the following lookup.
+           archive cid2
+           throw E
+         catch
+           E -> pure ()
+         (cid', _) <- fetchByKey @Keyed p
```

Comment:
This is the only test that currently fails. As described in the comments, my intuition is that consuming choices with a key basically have an implicit `fetchByKey` before (ignoring the visibility check that we don’t have for the exercise). That is a read so it isn’t rolled back so therefore the test should succeed.

If that’s not the case, we need semantics that are quite different from the current implementation.

Reply:
To make that example even more annoying, consider that we don’t know if the contract we see via the archive passes the visibility check for contract keys! This could mean that the `fetchByKey` has to fail here. instead of returning `cid2` which seems very confusing.

Reply:
One way out of this mess could be to limit the effect of consuming choices to those cases:
1. Set `keys` to `None` if it was `Some(cid)` before and the choice was exercised on cid.
2. Set `keys` to `None` if there was no entry in `keys`.

Only seems to be getting more convoluted though.

Reply:
What about the alternative where we do not update the key-map on consumption, but instead check for non-consumedness on read from the key-map?

Reply:
As discussed, this is what we’re going to do but not in this PR (and ignoring effects for multi-keys).